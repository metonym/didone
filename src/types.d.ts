export type DotEnvValue = {
  /**
   * The key of the environment variable.
   */
  key: string;

  /**
   * The value of the environment variable.
   * Multiline values are supported.
   */
  value: string;

  /**
   * `true` if the key is a duplicate of another key.
   * Its value will still be parsed.
   */
  duplicate: boolean;
};
