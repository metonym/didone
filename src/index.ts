export { parse } from "./parse";
export { serialize } from "./serialize";

