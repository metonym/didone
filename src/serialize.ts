import type { DotEnvValue } from "./types";

type SerializeOptions = {
  /**
   * If `true`, duplicate keys will be removed.
   * @default false
   */
  removeDuplicates?: boolean;
};

export function serialize(values: DotEnvValue[], options?: SerializeOptions) {
  const { removeDuplicates = false } = options || {};

  let text = "";

  for (const { key, value, duplicate } of values) {
    if (removeDuplicates === true && duplicate) continue;
    text += `${key}=${/(#|\n)/.test(value) ? `"${value}"` : value}\n`;
  }

  return text.trimEnd();
}
