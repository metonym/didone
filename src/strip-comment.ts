/**
 * Removes inline comments from a single-line string.
 * "#" within a quoted string should be preserved.
 * @example `foo # bar` -> "foo"
 * @example `"foo # bar"` -> "foo # bar"
 */
export function stripComment(text: string) {
  const chars = text.split("");
  const start = chars[0] === '"' ? 1 : 0;
  const break_char = chars[0] === '"' ? '"' : "#";

  let value = "";

  for (let i = start; i < chars.length; i++) {
    const char = chars[i];

    if (char === break_char) break;
    value += char;
  }

  return value.trim();
}
