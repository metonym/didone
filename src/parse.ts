import { stripComment, stripInlineComment } from "./strip-comment";
import type { DotEnvValue } from "./types";

const RE_SET_VERB = /^(export|set)\s+/;
const RE_WHITE_SPACE = /\s+/g;
const RE_BOUNDING_QUOTES_DOUBLE = /^"|"$/g;
const RE_BOUNDING_QUOTES_SINGLE = /^'|'$/g;
const RE_QUOTE_WITH_COMMENT = /(('.*?'|".*?")(\s*#\s*.*)?)$/;
const RE_QUOTE = /("|')/;

export function parse(text: string) {
  const lines = text.trim().split("\n");
  const keys = new Set<DotEnvValue["key"]>();
  const values = new Set<DotEnvValue>();

  for (let i = 0; i < lines.length; i++) {
    const line = lines[i].trim();
    if (line.startsWith("#")) continue;
    if (!line.includes("=")) continue;

    const [k, ...v] = line.split("=");
    const key = k.replace(RE_SET_VERB, "").trim();

    if (!key) continue;
    if (RE_WHITE_SPACE.test(key)) continue;

    let value = v.join("=").trim();

    if (RE_QUOTE.test(value[0]) && !RE_QUOTE_WITH_COMMENT.test(value)) {
      const start_quote = value[0];
      const quote_bound =
        start_quote === '"'
          ? RE_BOUNDING_QUOTES_DOUBLE
          : RE_BOUNDING_QUOTES_SINGLE;
      const quote_end = new RegExp(start_quote + "$");
      let multiline_value = value.replace(quote_bound, "");

      while (i < lines.length - 1) {
        i += 1;

        if (
          lines[i].includes(start_quote) &&
          !/\\/.test(lines[i].charAt(lines[i].indexOf(start_quote) - 1))
        ) {
          multiline_value +=
            "\n" +
            stripInlineComment(lines[i], start_quote).replace(quote_end, "");
          break;
        }

        multiline_value += "\n" + stripComment(lines[i]).replace(quote_end, "");
      }

      value = multiline_value;
    } else {
      value = stripComment(value);
    }

    values.add({ key, value, duplicate: keys.has(key) });
    keys.add(key);
  }

  return Array.from(values);
}
