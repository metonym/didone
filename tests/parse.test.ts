import { describe, expect, test } from "bun:test";
import { parse } from "../src";

describe("parse", () => {
  test.each(["1=bar", "Ü="])("invalid key", (text) => {
    expect(parse(text)).toEqual([]);
  });

  test.each([
    "export FOO=bar",
    "FOO=bar",
    "FOO=bar # comment",
    " FOO= bar ",
    "\nFOO=bar\n ",
    '# Comment\nFOO="bar"\n\n# Another comment',
  ])("single-line", (text) => {
    expect(parse(text)).toEqual([
      {
        duplicate: false,
        key: "FOO",
        value: "bar",
      },
    ]);
  });

  test.each([
    "export foo=bar",
    "foo=bar",
    " foo= bar ",
    "foo=bar # comment",
    "\nfoo=bar\n ",
    '# Comment\nfoo="bar"\n\n# Another comment',
  ])("single-line, lowercase", (text) => {
    expect(parse(text)).toEqual([
      {
        duplicate: false,
        key: "foo",
        value: "bar",
      },
    ]);
  });

  test.each([
    'FOO="abc#xyz"',
    'FOO="abc#xyz" # comment',
    '# comment\nFOO="abc#xyz"\n',
  ])("single-line, quoted", (text) => {
    expect(parse(text)).toEqual([
      {
        duplicate: false,
        key: "FOO",
        value: "abc#xyz",
      },
    ]);
  });

  test("multiline with inline comment", () => {
    expect(parse(`FOO="bar\nbaz"    # comment`)).toEqual([
      {
        duplicate: false,
        key: "FOO",
        value: "bar\nbaz",
      },
    ]);
  });

  test("multiline with inline comment", () => {
    expect(
      parse(`FOO="start
...
end"    # comment`)
    ).toEqual([
      {
        duplicate: false,
        key: "FOO",
        value: "start\n...\nend",
      },
    ]);
  });

  test("multiline with comments", () => {
    expect(
      parse(`# Comment
FOO="start"

## Another comment
`)
    ).toEqual([
      {
        duplicate: false,
        key: "FOO",
        value: "start",
      },
    ]);
  });

  test("non-env var text", () => {
    expect(
      parse(`type DotEnvValue = {
  key: string;
  value: string;
  duplicate: boolean;
};
`)
    ).toEqual([]);
  });

  test("unterminated quoted string", () => {
    expect(
      parse(`APP_ENV="development
A_SECRET=123456789abcdef`)
    ).toEqual([
      {
        duplicate: false,
        key: "APP_ENV",
        value: "development\nA_SECRET=123456789abcdef",
      },
    ]);
  });

  test("kitchen sink", () => {
    expect(
      parse(`# Database configuration
DB_HOST=localhost
DB_PORT="5432" # Port is read as a string
DB_USER=myuser`)
    ).toMatchSnapshot();

    expect(
      parse(`
# comment
export SMTP=4000
foo=bar
FOO=42
WITH-HYPEHN=1
WITH.period=1
ÜBER=ÜBER
234=123
SIMPLE=xyz123
WITH_INLINE_COMMENT=abc # comment
WITH_INLINE_COMMENT2="abc"
SINGLE_WITH_LINE_BREAKS="Multiple\nLines and variable substitution"
PRIVATE_KEY="-----BEGIN RSA PRIVATE KEY-----
...
Kh9NV...

...
-----END RSA PRIVATE KEY-----
"
SIMPLE=abc
PRIVATE_KEY=abc
`)
    ).toMatchSnapshot();
  });
});
