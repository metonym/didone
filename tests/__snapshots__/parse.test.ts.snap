// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`parse kitchen sink 1`] = `
[
  {
    "duplicate": false,
    "key": "DB_HOST",
    "value": "localhost",
  },
  {
    "duplicate": false,
    "key": "DB_PORT",
    "value": "5432",
  },
  {
    "duplicate": false,
    "key": "DB_USER",
    "value": "myuser",
  },
]
`;

exports[`parse kitchen sink 2`] = `
[
  {
    "duplicate": false,
    "key": "SMTP",
    "value": "4000",
  },
  {
    "duplicate": false,
    "key": "foo",
    "value": "bar",
  },
  {
    "duplicate": false,
    "key": "FOO",
    "value": "42",
  },
  {
    "duplicate": false,
    "key": "WITH-HYPEHN",
    "value": "1",
  },
  {
    "duplicate": false,
    "key": "WITH.period",
    "value": "1",
  },
  {
    "duplicate": false,
    "key": "ÜBER",
    "value": "ÜBER",
  },
  {
    "duplicate": false,
    "key": "234",
    "value": "123",
  },
  {
    "duplicate": false,
    "key": "SIMPLE",
    "value": "xyz123",
  },
  {
    "duplicate": false,
    "key": "WITH_INLINE_COMMENT",
    "value": "abc",
  },
  {
    "duplicate": false,
    "key": "WITH_INLINE_COMMENT2",
    "value": "abc",
  },
  {
    "duplicate": false,
    "key": "SINGLE_WITH_LINE_BREAKS",
    "value": 
"Multiple
Lines and variable substitution"
,
  },
  {
    "duplicate": false,
    "key": "PRIVATE_KEY",
    "value": 
"-----BEGIN RSA PRIVATE KEY-----
...
Kh9NV...

...
-----END RSA PRIVATE KEY-----
"
,
  },
  {
    "duplicate": true,
    "key": "SIMPLE",
    "value": "abc",
  },
  {
    "duplicate": true,
    "key": "PRIVATE_KEY",
    "value": "abc",
  },
  {
    "duplicate": true,
    "key": "PRIVATE_KEY",
    "value": "abc",
  },
  {
    "duplicate": false,
    "key": "PRIVATE_KEY_SINGLE",
    "value": 
"-----BEGIN RSA PRIVATE KEY-----
...
-----END RSA PRIVATE KEY-----"
,
  },
]
`;
